### PR TITLE
Typha migration

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

For Kubernetes API datastore clusters, Invoke migration from Typha.

Continuation of #93 by @robbrockbank.  I took his work and reworked it a bit to share the V3 client between the main Typha code and the migrator.  I think that closes some corner cases and avoids leaking a client, which we have no method to fully close.

To do that, I had to make the test shims a bit less nice (the datastore shim now exposes the full client API so that it can be passed into the migrator).

## Todos
- [ ] Unit tests (full coverage)
- [ ] Integration tests (delete as appropriate) In plan/Not needed/Done
- [ ] Documentation
- [ ] Backport
- [ ] Release note

## Release Note

```release-note
None required
```

### CODE REVIEW

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -177,20 +184,21 @@ configRetry:
 			continue configRetry
 		}
 
-		// We should now have enough config to connect to the datastore.
-		datastoreConfig := configParams.DatastoreConfig()
-
-		t.DatastoreClient, err = t.NewClientV3(datastoreConfig)
+		// Validate the config params
+		err = configParams.Validate()
```
Comment (neiljerram) [inline pkg/daemon/daemon.go]: Why has this validate check moved from after merging datastore config, to before?

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -141,14 +143,19 @@ func (t *TyphaDaemon) ParseCommandLineArgs(argv []string) {
...
+func (t *TyphaDaemon) LoadConfiguration(ctx context.Context) error {
 	// Load the configuration from all the different sources including the
 	// datastore and merge. Keep retrying on failure.  We'll sit in this
 	// loop until the datastore is ready.
 	log.Infof("Loading configuration...")
 	var configParams *config.Config
+	var datastoreConfig apiconfig.CalicoAPIConfig
 configRetry:
 	for {
+		if ctx.Err() != nil {
```
[the remark this answers is not in the document]
Comment (fasaxc) [reply pkg/daemon/daemon.go]: Agreed, was thinking about making that change too

File: pkg/daemon/daemon.go (hunk 1 above)
Comment (fasaxc) [reply pkg/daemon/daemon.go]: I realised that doing it after didn't make sense for Typha, which doesn't load its config from the datastore.
